core/src/main/java/com/linecorp/armeria/common/RequestContext.java

```diff
@@ -276,6 +209,114 @@ default EventLoop contextAwareEventLoop() {
+     * <pre>{@code
+     * try (SafeCloseable ignored = ctx.push()) {
+     *     ...
+     * }
+     * }</pre>
+     *
+     * <p>The callbacks added by {@link #onEnter(Consumer)} and {@link #onExit(Consumer)} will be invoked
+     * when the context is pushed to and removed from the thread-local stack respectively.
+     *
+     * <p>NOTE: In case of re-entrance, the callbacks will never run.
+     */
+    default SafeCloseable push() {
+        return push(true);
+    }
+
+    /**
+     * Pushes the specified context to the thread-local stack. To pop the context from the stack, call
+     * {@link SafeCloseable#close()}, which can be done using a {@code try-with-resources} block:
+     * <pre>{@code
+     * try (PushHandle ignored = ctx.push(true)) {
+     *     ...
+     * }
+     * }</pre>
+     *
+     * <p>NOTE: This method is only useful when it is undesirable to invoke the callbacks, such as replacing
+     *          the current context with another. Prefer {@link #push(RequestContext)} otherwise.
+     *
+     * @param runCallbacks if {@code true}, the callbacks added by {@link #onEnter(Consumer)} and
+     *                     {@link #onExit(Consumer)} will be invoked when the context is pushed to and
+     *                     removed from the thread-local stack respectively.
+     *                     If {@code false}, no callbacks will be executed.
+     *                     NOTE: In case of re-entrance, the callbacks will never run.
+     */
+    default SafeCloseable push(boolean runCallbacks) {
+        final RequestContext oldCtx = RequestContextThreadLocal.getAndSet(this);
+        if (oldCtx == this) {
+            // Reentrance
+            return () -> { /* no-op */ };
+        }
+
+        if (runCallbacks) {
+            if (oldCtx != null) {
+                oldCtx.invokeOnChildCallbacks(this);
+                invokeOnEnterCallbacks();
+                return () -> {
+                    invokeOnExitCallbacks();
+                    RequestContextThreadLocal.set(oldCtx);
+                };
+            } else {
+                invokeOnEnterCallbacks();
+                return () -> {
+                    invokeOnExitCallbacks();
+                    RequestContextThreadLocal.remove();
+                };
+            }
+        } else {
+            if (oldCtx != null) {
+                return () -> RequestContextThreadLocal.set(oldCtx);
+            } else {
+                return RequestContextThreadLocal::remove;
+            }
+        }
+    }
+
+    /**
+     * Pushes this context to the thread-local stack if there is no current context. If there is, this method
+     * will throw IllegalStateException. To pop the context from the stack, call {@link SafeCloseable#close()},
+     * which can be done using a {@code try-with-resources} block.
+     */
+    default SafeCloseable pushIfAbsent() {
+        return mapCurrent(currentContext -> {
```

Comment:
Now that this method is moved from the subclass, we don't need to use `mapCurrent()` anymore but use `RequestContextThreadLocal.get()` directly, which is supposed to be easier to read:

```java
final RequestContext oldCtx = RequestContextThreadLocal.getAndSet(this);
if (oldCtx != null && oldCtx != this) {
    throw new IllegalStateException(...);
}
return push();
```